eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelTypes.scala

```diff
@@ -105,34 +107,83 @@ case class BITCOIN_PARENT_TX_CONFIRMED(childTx: Transaction) extends BitcoinEven
        "Y8888P"   "Y88888P"  888       888 888       888 d88P     888 888    Y888 8888888P"   "Y8888P"
  */
 
-sealed trait Upstream
-object Upstream {
+/**
+ * Origin of a payment, answering both questions:
+ * - what actor in the app sent that htlc? (Origin.replyTo)
+ * - what are the parent(s) of this payment in the htlc chain? (Upstream)
+ */
+sealed trait Origin[+U <: Origin.Upstream] { def upstream: U }
```

Comment:
Here is my proposal:

```scala
  /**
   * Origin of a payment, answering both questions:
   * - what actor in the app sent that htlc? (Origin.replyTo)
   * - what are the upstream parent(s) of this payment in the htlc chain?
   */
  sealed trait Origin
  /** We haven't restarted since we sent the payment downstream: the origin actor is known. */
  sealed trait Hot extends Origin { def replyTo: ActorRef }
  /** We have restarted after the payment was sent, we have limited info and the origin actor doesn't exist anymore. */
  sealed trait Cold extends Origin

  /** The downstream HTLCs have been sent by a payment FSM and may be retried in case of failures. */
  sealed trait HasPaymentLifecycle extends Origin

  /** Our node is the origin of the payment. */
  sealed trait Local extends HasPaymentLifecycle { def id: UUID }
  case class LocalHot(replyTo: ActorRef, id: UUID) extends Local with Hot
  case class LocalCold(id: UUID) extends Local with Cold

  /** Our node forwarded a single incoming HTLC to an outgoing channel. */
  sealed trait ChannelRelayed extends Origin {
    def originChannelId: ByteVector32
    def originHtlcId: Long
    def amountIn: MilliSatoshi
    def amountOut: MilliSatoshi
  }
  case class ChannelRelayedHot(replyTo: ActorRef, add: UpdateAddHtlc, override val amountOut: MilliSatoshi) extends ChannelRelayed with Hot {
    override def originChannelId: ByteVector32 = add.channelId
    override def originHtlcId: Long = add.id
    override def amountIn: MilliSatoshi = add.amountMsat
  }
  case class ChannelRelayedCold(originChannelId: ByteVector32, originHtlcId: Long, amountIn: MilliSatoshi, amountOut: MilliSatoshi) extends ChannelRelayed with Cold

  /** Our node forwarded an incoming HTLC set to a remote outgoing node (potentially producing multiple downstream HTLCs).*/
  sealed trait TrampolineRelayed extends HasPaymentLifecycle { def htlcs: List[(ByteVector32, Long)] }
  case class TrampolineRelayedHot(replyTo: ActorRef, adds: Seq[UpdateAddHtlc]) extends TrampolineRelayed with Hot {
    override def htlcs: List[(ByteVector32, Long)] = adds.map(u => (u.channelId, u.id)).toList
    val amountIn: MilliSatoshi = adds.map(_.amountMsat).sum
    val expiryIn: CltvExpiry = adds.map(_.cltvExpiry).min
  }
  case class TrampolineRelayedCold(override val htlcs: List[(ByteVector32, Long)]) extends TrampolineRelayed with Cold
```

There are two distinct ways to then pattern match on `Origin`:
* separate them by Hot vs Cold (post-restart-handler handles only cold, relayers handle only hot and then pattern match further)
* separate them by type of relay (Local, ChannelRelayed, TrampolineRelayed) and then pattern match on hot vs cold

Reply:
Agreed, I'll try that right away.

Reply:
Wait, I remember now. There are cases where we know the `Upstream` but we don't yet have the sender. When sending trampoline payments if my memory serves right. Will still try though...

Reply:
If that's the case, maybe we can add a dedicated type but only for that specific case (instead of having every relay scenario have both a hot and cold upstream type)? I'm looking into it as well :)

Reply:
> Not sure where to put the new Upstream trait though.

I think that since it originates from `PaymentInitiator` (and only used for `SendPaymentConfig`) it's fine where you put it for now.

Reply:
Done with 4a25b39c4.